rpcs3/Emu/RSX/VK/VKTextureCache.h

```diff
@@ -198,21 +197,78 @@ namespace vk
+				vk::get_component_mapping(tex.format() & ~(CELL_GCM_TEXTURE_LN | CELL_GCM_TEXTURE_UN), tex.remap()),
+				vk::default_image_subresource_range());
+
+			// Upload all mipmap levels
+			const std::vector<rsx_subresource_layout> &subresources_layout = get_subresources_layout(tex);
+			u32 mipmap_level = 0;
+			u32 block_in_pixel = get_format_block_size_in_texel(format);
+			for (const rsx_subresource_layout &layout : subresources_layout)
+			{
+				// Needs to create an image per subresource in case linear resource doesn't support mipmap count > 1
+				m_texture_to_clean_after_sync.push_back(
+					std::make_unique<vk::image>(
+						*vk::get_current_renderer(), memory_type_mapping.host_visible_coherent,
+						VK_IMAGE_TYPE_2D,
+						vk_format,
+						layout.width_in_block * block_in_pixel, layout.height_in_block * block_in_pixel, 1, 1, 1, VK_SAMPLE_COUNT_1_BIT, VK_IMAGE_LAYOUT_PREINITIALIZED,
+						VK_IMAGE_TILING_LINEAR, VK_IMAGE_USAGE_TRANSFER_SRC_BIT, 0));
+				vk::image *staging_image = m_texture_to_clean_after_sync.back().get();
+
+				void *data;
+				CHECK_RESULT(vkMapMemory(*vk::get_current_renderer(), staging_image->memory->memory, 0, staging_image->memory->info.allocationSize, 0, (void**)&data));
+				gsl::span<gsl::byte> mapped{ (gsl::byte*)data, gsl::narrow<int>(staging_image->memory->info.allocationSize) };
+
+				VkImageSubresource subres = {};
+				subres.aspectMask = VK_IMAGE_ASPECT_COLOR_BIT;
+				subres.mipLevel = mipmap_level;
+				subres.arrayLayer = 0;
+
+				VkSubresourceLayout subres_layout;
+
+				vkGetImageSubresourceLayout(*vk::get_current_renderer(), staging_image->value, &subres, &subres_layout);
+				upload_texture_subresource(mapped.subspan(subres_layout.offset), layout, tex.format() & ~(CELL_GCM_TEXTURE_LN | CELL_GCM_TEXTURE_UN), !(tex.format() & CELL_GCM_TEXTURE_LN), subres_layout.rowPitch);
+
+				vkUnmapMemory(*vk::get_current_renderer(), staging_image->memory->memory);
+
+				change_image_layout(cmd, staging_image->value, VK_IMAGE_LAYOUT_PREINITIALIZED, VK_IMAGE_LAYOUT_TRANSFER_SRC_OPTIMAL, VK_IMAGE_ASPECT_COLOR_BIT);
+
+				VkImageCopy rgn = {};
+				rgn.extent.depth = 1;
+				rgn.extent.width = layout.width_in_block * block_in_pixel;
+				rgn.extent.height = layout.height_in_block * block_in_pixel;
+				rgn.dstOffset = { 0, 0, 0 };
+				rgn.srcOffset = { 0, 0, 0 };
+				rgn.srcSubresource = { VK_IMAGE_ASPECT_COLOR_BIT, 0, 0, 1 };
+				rgn.dstSubresource = { VK_IMAGE_ASPECT_COLOR_BIT, mipmap_level, 0, 1 };
+
+				vkCmdCopyImage(cmd, staging_image->value, VK_IMAGE_LAYOUT_TRANSFER_SRC_OPTIMAL, cto.uploaded_texture->value, VK_IMAGE_LAYOUT_TRANSFER_DST_OPTIMAL, 1, &rgn);
+
+				mipmap_level++;
+			}
+
+			change_image_layout(cmd, cto.uploaded_texture->value, VK_IMAGE_LAYOUT_TRANSFER_DST_OPTIMAL, VK_IMAGE_LAYOUT_SHADER_READ_ONLY_OPTIMAL, VK_IMAGE_ASPECT_COLOR_BIT);
```

Comment:
The purpose of vk::texture::init was largely to wrap the block above. Its very verbose and it would be easy/neat to just have a function call instead in my opinion.